src/irmin/inode.ml

```diff
@@ -0,0 +1,983 @@
+
+    let empty : 'a. 'a layout -> 'a t =
+     fun _ ->
+      let hash = lazy (Node.hash Node.empty) in
+      { stable = true; hash; v = Values StepMap.empty }
+
+    let values layout vs =
+      let length = StepMap.cardinal vs in
+      if length = 0 then empty layout
+      else
+        let v = Values vs in
+        let hash = lazy (Bin.V.hash (to_bin_v layout v)) in
+        { hash; stable = false; v }
+
+    let tree layout is =
+      let v = Tree is in
+      let hash = lazy (Bin.V.hash (to_bin_v layout v)) in
+      { hash; stable = false; v }
+
+    let of_values layout l = values layout (StepMap.of_list l)
+
+    let is_empty t =
+      match t.v with Values vs -> StepMap.is_empty vs | Tree _ -> false
+
+    let find_value layout ~depth t s =
+      let target_of_ptr = Ptr.target layout in
+      let rec aux ~depth = function
+        | Values vs -> ( try Some (StepMap.find s vs) with Not_found -> None)
+        | Tree t -> (
+            let i = index ~depth s in
+            let x = t.entries.(i) in
+            match x with
+            | None -> None
+            | Some i -> aux ~depth:(depth + 1) (target_of_ptr i).v)
+      in
+      aux ~depth t.v
+
+    let find layout t s = find_value ~depth:0 layout t s
+
+    let rec add layout ~depth ~copy ~replace t s v k =
+      match t.v with
+      | Values vs ->
+          let length =
+            if replace then StepMap.cardinal vs else StepMap.cardinal vs + 1
+          in
+          let t =
+            if length <= Conf.entries then values layout (StepMap.add s v vs)
+            else
+              let vs = StepMap.bindings (StepMap.add s v vs) in
+              let empty =
+                tree layout
+                  { length = 0; depth; entries = Array.make Conf.entries None }
+              in
+              let aux t (s, v) =
+                (add [@tailcall]) layout ~depth ~copy:false ~replace t s v
+                  (fun x -> x)
+              in
+              List.fold_left aux empty vs
+          in
+          k t
+      | Tree t -> (
+          let length = if replace then t.length else t.length + 1 in
+          let entries = if copy then Array.copy t.entries else t.entries in
+          let i = index ~depth s in
+          match entries.(i) with
+          | None ->
+              let target = values layout (StepMap.singleton s v) in
+              entries.(i) <- Some (Ptr.of_target layout target);
+              let t = tree layout { depth; length; entries } in
+              k t
+          | Some n ->
+              let t = Ptr.target layout n in
+              add layout ~depth:(depth + 1) ~copy ~replace t s v
+              @@ fun target ->
+              entries.(i) <- Some (Ptr.of_target layout target);
+              let t = tree layout { depth; length; entries } in
+              k t)
+
+    let add layout ~copy t s v =
+      (* XXX: [find_value ~depth:42] should break the unit tests. It doesn't. *)
```

Comment:
is the `XXX` remark still true?

Reply:
I think so

Reply:
Do you know why it doesn't break them? That seems rather odd!

Reply:
Nope :)